src/shared_queue.cc

```diff
@@ -21,23 +21,72 @@
 #include <unordered_map>
 #include <unordered_set>
 
+#include "include/proxy-wasm/vm_id_handle.h"
+
 namespace proxy_wasm {
 
 SharedQueue &getGlobalSharedQueue() {
   static auto *ptr = new SharedQueue;
   return *ptr;
 }
 
+SharedQueue::SharedQueue(bool register_vm_id_callback) {
+  if (register_vm_id_callback) {
+    registerVmIdHandleCallback([this](std::string_view vm_id) { this->deleteByVmId(vm_id); });
+  }
+}
+
+void SharedQueue::deleteByVmId(std::string_view vm_id) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto queue_keys = vm_queue_keys_.find(std::string(vm_id));
+  if (queue_keys != vm_queue_keys_.end()) {
+    for (std::pair<std::string, std::string> queue_key : queue_keys->second) {
+      auto token = queue_tokens_.find(queue_key);
+      if (token != queue_tokens_.end()) {
+        queues_.erase(token->second);
+        queue_tokens_.erase(token);
+      }
+    }
+    vm_queue_keys_.erase(queue_keys);
+  }
+}
+
+uint32_t SharedQueue::nextQueueToken() {
+  // TODO(@mathetake): Should we handle the case where the queue overflows, i.e. the number of used
+  // tokens exceeds the max of uint32? If it overflows, the following loop never exits.
+  while (true) {
+    uint32_t token = next_queue_token_++;
+    if (token == 0) {
+      continue; // 0 is an illegal token.
+    }
+
+    if (queues_.find(token) == queues_.end()) {
+      return token;
+    }
+  }
+}
+
 uint32_t SharedQueue::registerQueue(std::string_view vm_id, std::string_view queue_name,
                                     uint32_t context_id, CallOnThreadFunction call_on_thread,
                                     std::string_view vm_key) {
   std::lock_guard<std::mutex> lock(mutex_);
-  auto key = std::make_pair(std::string(vm_id), std::string(queue_name));
+  std::string vid = std::string(vm_id);
+  auto key = std::make_pair(vid, std::string(queue_name));
+
+  QueueKeySet *queue_keys;
+  auto map_it = vm_queue_keys_.find(vid);
+  if (map_it == vm_queue_keys_.end()) {
+    queue_keys = &vm_queue_keys_[vid];
+  } else {
+    queue_keys = &map_it->second;
+  }
+  queue_keys->insert(key);
+
   auto it = queue_tokens_.insert(std::make_pair(key, static_cast<uint32_t>(0)));
   if (it.second) {
     it.first->second = nextQueueToken();
-    queue_token_set_.insert(it.first->second);
   }
```

Comment:
Instead of searching `vm_queue_keys_` first, you could move that code into the `if (it.second) {` block, since we know that the `vm_queue_keys_` already exists if it evaluates to `false`.

Reply:
right. Thanks